apiserver/oauth-proxy-proxy.md

```diff
@@ -0,0 +1,141 @@
+
+### Non-Goals
+
+1. Making generic code to rewrite proxy transport configuration.
+
+## Proposal
+
+The `oauth-proxy` can be written to use targeted list/watch against the proxy.config.openshift.io and 
+`oc get -n openshift-config-managed configmaps/proxy-ca` to build a proxy transport.  Using standard informers with single item
+watches allows RBAC to only expose GET for individual items.  It also means that an efficient swap mechanism can be built
+to ensure performant behavior.
+
+This is an intricate thing to write, so we avoided it for every other proxy component, but this one is a component that is
+re-used by numerous external teams by directly embedding it into a pod spec.  This changes the tradeoff, so it becomes worthwhile
+to write this code.
+
+### User Stories [optional]
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. If HTTP env vars are set, the dynamic proxy-support code must be disabled.  The user is in charge of what the `oauth-proxy`
+ does in these cases.
+2. The list/watch driving the informers must be focused to a single item, so the RBAC special case for single item watch
+ takes effect.
+3. To be performant, the proxy and TLS configuration should be cached in something like a `sync.Atomic`.
+
+
+### Risks and Mitigations
+
+This code is intricate to write, but luckily our monitoring stack already relies on this.  If their tests pass when they
+update to use the feature, the feature works.
+
+## Design Details
+
+### Test Plan
+
+We would get @sur to update the monitoring stack to avoid setting the HTTP env vars and ensure that the proxy e2e test
+still works.
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
+
+None, fully compatible.
+
+### Version Skew Strategy
+
+None, fully compatible.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
+
+1. Using the internal service DNS name. The .well-known endpoint is public and conformant to a standard RFC.
+ It can only include a single URL, which means it must
+ be public so that external clients like `oc`, the webconsole, and anything else can successfully use it.  Using an internal
+ service name would also preclude the kube-apiserver from ever using an external oauth provider like keycloak.
+
+2. Setting HTTP env vars.  We are doing this is to deliver a mostly working 4.2.0, but has undesireable, long-term characteristics.
+ Setting env vars on the `oauth-proxy` container is possible, but it requires informing every team using the `oauth-proxy`
+ that they need to change.  If they are set, we should honor them, but communicating the scope of this change is difficult; 
+ we can do it for the few in our payload before we ship 4.2.0, but the fanout afterwards is hard.
```

Comment:
it's also insufficient, right?  because we still need the CA (for MITM proxies), not just the env vars.

Reply:
> it's also insufficient, right? because we still need the CA (for MITM proxies), not just the env vars.

I don't think containing a MITM option for oauth-proxy in 4.2.0 is a practical choice.

Reply:
agreed it's out of scope for 4.2.0, but i assumed all of this was for 4.3 (and then potentially backporting to 4.2.z)?

Reply:
meaning this alternative (setting env vars), while the only reasonable thing we can do for 4.2.0, is not a viable alternative in general(long term) not only for the reasons you cited, but also because it doesn't address MITM CAs.  (is that i was trying to say).